src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WorkspaceDiagnosticsRefresher.cs

```diff
@@ -0,0 +1,118 @@
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer;
+
+internal sealed class WorkspaceDiagnosticsRefresher : IRazorStartupService, IDisposable
+{
+    private readonly AsyncBatchingWorkQueue _queue;
+    private readonly IProjectSnapshotManager _projectSnapshotManager;
+    private readonly IClientCapabilitiesService _clientCapabilitiesService;
+    private readonly IClientConnection _clientConnection;
+    private bool? _supported;
+    private CancellationTokenSource _disposeTokenSource = new();
+
+    public WorkspaceDiagnosticsRefresher(
+        IProjectSnapshotManager projectSnapshotManager,
+        IClientCapabilitiesService clientCapabilitiesService,
+        IClientConnection clientConnection,
+        TimeSpan? delay = null)
+    {
+        _clientConnection = clientConnection;
+        _projectSnapshotManager = projectSnapshotManager;
+        _clientCapabilitiesService = clientCapabilitiesService;
+        _queue = new(
+            delay ?? TimeSpan.FromMilliseconds(200),
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+        _projectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+    }
+
+    public void Dispose()
+    {
+        if (_disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+        _projectSnapshotManager.Changed -= ProjectSnapshotManager_Changed;
+    }
+
+    private ValueTask ProcessBatchAsync(CancellationToken token)
+    {
+        _clientConnection
+            .SendNotificationAsync(Methods.WorkspaceDiagnosticRefreshName, token)
+            .Forget();
+
+        return default;
+    }
+
+    private void ProjectSnapshotManager_Changed(object? sender, ProjectChangeEventArgs e)
+    {
+        if (e.SolutionIsClosing)
+        {
+            return;
+        }
+
+        if ( _disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _supported ??= GetSupported();
+
+        if (_supported != true)
+        {
+            return;
+        }
+
+        if (e.Kind is not ProjectChangeKind.DocumentChanged)
+        {
+            _queue.AddWork();
+        }
+    }
+
+    private bool? GetSupported()
+    {
+        if (!_clientCapabilitiesService.CanGetClientCapabilities)
+        {
+            return null;
+        }
+
+        return _clientCapabilitiesService.ClientCapabilities.Workspace?.Diagnostics?.RefreshSupport;
+    }
+
+    internal TestAccessor GetTestAccessor()
+        => new(this);
+
+    internal sealed class TestAccessor
+    {
+        private readonly WorkspaceDiagnosticsRefresher _instance;
+
+        public TestAccessor(WorkspaceDiagnosticsRefresher instance)
+        {
+            _instance = instance;
+        }
+
+        public Task WaitForRefreshAsync()
+        {
+            return _instance._queue.WaitUntilCurrentBatchCompletesAsync();
```

Comment:
FWIW, there might be a hang in tests where `WaitForRefreshAsync()` is called after the `WorkspaceDiagnosticRefresher` is dispsoed.

Reply:
Is it better to check disposal here and return a `Task.CompletedTask`?

Reply:
I quick fix might be something like...

```C#
if (_instance._disposeTokenSource.IsCancellationRequested)
{
    return Task.CompletedTask;
}
```

Reply:
> Is it better to check disposal here and return a `Task.CompletedTask`?

LOL. I didn't refresh quickly enough to see your comment. 🤣